lib/active_encode/engine_adapters/ffmpeg_adapter.rb

```diff
@@ -0,0 +1,185 @@
+require 'fileutils'
+require 'nokogiri'
+
+module ActiveEncode
+  module EngineAdapters
+    class FfmpegAdapter
+      WORK_DIR = ENV["ENCODE_WORK_DIR"] || "encodes" # Should read from config
+
+      def create(input_url, options = {})
+        new_encode = ActiveEncode::Base.new(input_url, options)
+        new_encode.id = SecureRandom.uuid
+        new_encode.state = :running
+        new_encode.current_operations = []
+        new_encode.percent_complete = 10
+        new_encode.errors = []
+        new_encode.created_at = Time.new
+        new_encode.updated_at = Time.new
+
+        new_encode.input.url = input_url
+        new_encode.input.created_at = new_encode.created_at
+        new_encode.input.updated_at = new_encode.created_at
+
+        new_encode.output = []
+
+        # Create a working directory that holds all output files related to the encode
+        FileUtils.mkdir_p working_path("", new_encode.id)
+
+        # Extract technical metadata from input file
+        `mediainfo --Output=XML --LogFile=#{working_path("input_metadata", new_encode.id)} #{input_url}`
+        new_encode.input.assign_tech_metadata get_tech_metadata(working_path("input_metadata", new_encode.id))
+
+        # Run the ffmpeg command and save its pid
+        command = ffmpeg_command(new_encode, options)
+        pid = Process.spawn(command)
+        File.open(working_path("pid", new_encode.id), 'w') { |file| file.write pid }
+        new_encode.input.id = pid
+
+        # Prevent zombie process
+        Process.detach(pid)
+
+        new_encode
+      end
+
+      # Return encode object from file system
+      def find(id, opts={})
+        encode_class = opts[:cast]
+        encode = ActiveEncode::Base.new(nil, opts)
+        encode.id = id
+        encode.errors = []
+        encode.current_operations = []
+        encode.created_at = File.mtime working_path("pid", id)
+        encode.updated_at = File.mtime working_path("progress", id)
+
+        pid = get_pid(id)
+        encode.input.id = pid
+        encode.input.url = "N/A"
+        encode.input.created_at = encode.created_at
+        encode.input.updated_at = encode.updated_at
+        encode.input.assign_tech_metadata get_tech_metadata(working_path("input_metadata", encode.id))
+
+        # Read progress data from file
+        progress_data = File.read working_path("progress", id)
+        encode.percent_complete = calculate_percent_complete encode, progress_data
+
+        if running? pid
+          encode.state = :running
+          encode.current_operations = ["transcoding"]
+        else
+          error = File.read working_path("error.log", id)
+          if error.present?
+            encode.state = :failed
+            encode.errors = [error]
+          elsif progress_ended?(progress_data) && encode.percent_complete == 100
+            encode.state = :completed
+          elsif encode.percent_complete < 100
```

Comment:
is there a case when progress is ended (shown by ffmpeg log) but percent_complete is not 100? if so, what status shall the encode be? (error?) in case the job is cancelled, the progress_ended? will still return false, right?